elastic/apm-server
# Always set timestamp on APMEvents for incoming http requests

## Motivation/summary
APM Server does not record the timestamp from incoming RUM events when `capture_personal_data: false`. This PR ensures the timestamp is always set in the initial `APMEvent`. 
The approval tests are ignoring the concrete timestamp, as it is dynamic and would be ever changing. The PR adds functionality allowing to implement a custom check for dynamic attributes before dumping and comparing them to the approved documents. It is used to ensure that the `@timestamp` is not zero, without checking for the actual value. 

## Checklist


## How to test these changes

### manual test
* spin up apm-server `>= 8.0`, enable rum and configure `apm-server.capture-personal-data: false` (standalone or managed mode doesn't matter)
* send a rum request to the apm-server 
* observe that the created `@timestamp` is a valid timestamp

When the same test is run on current `main` branch (without this fix) the `@timestamp` value would be the `zero` value. 

### system test
Without the fix, the newly introduced test was failing like this
```
2022/03/17 18:10:53 Built /Users/simitt/coding/go/elastic/apm-server/apm-server
--- FAIL: TestRUMCaptureNoPersonalData (5.79s)
    rum_test.go:255: calling Check() for dynamic field @timestamp failed
    server.go:191: log file: /Users/simitt/coding/go/elastic/apm-server/systemtest/logs/TestRUMCaptureNoPersonalData/apm-server
FAIL
FAIL	github.com/elastic/apm-server/systemtest	30.525s
```

## Related issues
fixes #7566
